core/src/main/scala/org/apache/spark/SSLOptions.scala

```diff
@@ -102,6 +106,36 @@ private[spark] case class SSLOptions(
     }
   }
 
+  /*
+   * The supportedAlgorithms set is a subset of the enabledAlgorithms that
+   * are supported by the current Java security provider for this protocol.
+   */
+  private val supportedAlgorithms: Set[String] = {
+    var context: Option[SSLContext] = Some(SSLContext.getDefault)
```

Comment:
All you are missing is my inexperience in Scala!  So how about something like this?
```scala
    var context: SSLContext = null
    try {
      context = SSLContext.getInstance(protocol.orNull)
      context.init(null, null, null)
    } catch {
      case npe: NullPointerException =>
        logDebug("No SSL protocol specified")
        context = SSLContext.getDefault
      case nsa: NoSuchAlgorithmException =>
        logDebug(s"No support for requested SSL protocol ${protocol.get}")
        context = SSLContext.getDefault
    }

    val providerAlgorithms = context.getServerSocketFactory.getSupportedCipherSuites.toSet
```